### DISCUSSION TITLE
Expose ic-wasm interface from dfx

### DISCUSSION THREAD

**Discussion by viviveevee** [dfinity/sdk#2734] [Feature Requests]
@lastmjs requests this [here](https://forum.dfinity.org/t/wasm-binary-optimization-in-dfx/16322).

> I’ve been using ic-cdk-optimizer and am attempting to switch over to ic-wasm. These tools are essential for Azle and Kybra, the TypeScript and Python CDKs. One issue of complication that arises when developers try to use Azle and Kybra, is that each CDK must install ic-cdk-optimizer or ic-wasm for the developer, or expect them to install these tools. Requiring any kind of Rust tooling for TypeScript and Python devs is undesirable.

> Basically, can we incorporate ic-wasm into the dfx build process somehow, especially so that developers don’t have to independently install ic-wasm?

**Comment by smallstepman** [comment]
I wonder how hard would it be to compile ic-wasm to dynlib or wasm, so that it will be possible to run it from other languages and reuse the current codebase at the same time

**Reply by viviveevee** [reply]
Interesting, but not really relevant to this discussion I think. For this it'd be enough to have it available under `$(dfx cache show)/ic-wasm` or something like that just so other tools don't have to download it separately

**Reply by smallstepman** [reply]
TIL

**Comment by lwshang** [comment]
Since this [merged PR](https://github.com/dfinity/sdk/pull/2578), `dfx` default to run `ic-wasm shrink` pass during canister build. No need to install `ic-wasm` binary.

This feather was included since `dfx v0.12.0-beta.3`.

**Comment by lastmjs** [comment]
If I understand @lwshang correctly, once `dfx 12` is released this feature request should be fully implemented. Correct?

**Reply by viviveevee** [reply]
Your call, Jordan. I think the currently important parts (shrinking and metadata manipulation) are covered with dfx 0.12, but the utility itself is not reachable
